### DISCUSSION TITLE
Mapbox inegration: change cursor on hover

### DISCUSSION THREAD

**Discussion by GarrettCannon** [visgl/deck.gl#5893] [Q&A] [answered]
Hi, I am integrating Deck.gl with an existing Mapbox application I built previously. Following the examples in the docs I can add the layer and it seems to work fine with the test data I have. For some of the Mapbox layers not managed through Deck.gl, I was changing the cursor to a pointer when the cursor was over certain features. However with Deck.gl added the cursor never changes, the event handler is running, but it seems like Deck.gl is immediately switching the cursor back to a grab/grabbing before the pointer cursor gets rendered.

Is there a way to turn this behavior off or work around it?

Edit:

Also here are two gifs

Without Deck.gl
![mapbox-no-deckgl](https://user-images.githubusercontent.com/44881574/122616058-a68e2e00-d057-11eb-8187-a321e1d7ecf4.gif)

With Deck.gl
![mb-with-deckgl](https://user-images.githubusercontent.com/44881574/122616080-b574e080-d057-11eb-8707-6c591287adf3.gif)

**Comment by Pessimistress** [comment, marked as answer]
Can you share your code?

**Reply by GarrettCannon** [reply]
Map Init
```
new mapboxgl.Map({
        container: "map",
        style: import.meta.env.VITE_MAPBOX_STYLES,
        accessToken: import.meta.env.VITE_ACCESS_TOKEN,
        logoPosition: "bottom-left",
        fadeDuration: 100,
        trackResize: true,
        bounds: new mapboxgl.LngLatBounds([123, 85, -230, -70]),
        maxBounds: new mapboxgl.LngLatBounds([-600, -70, 600, 85]),
        maxZoom: 8,
        bearingSnap: 360,
        pitchWithRotate: false,
        dragRotate: false,
        touchPitch: false,
        doubleClickZoom: false,
        renderWorldCopies: renderWorldCopies,
        antialias: true,
      });
```
Deck.Gl
```
//Not exposed by type signatures but does exist
//@ts-ignore
const gl = Map.current?.painter.context.gl;

DeckGl.current = new Deck({
  gl: gl,
  getCursor: () => "",
  layers: [
    new ArcLayer({
      id: "connections-deck-gl",
      pickable: false,
      data: [{ start: [0, 0], end: [-85, 40] }],
      getSourcePosition: (d: any) => d.start,
      getTargetPosition: (d: any) => d.end,
      getColor: [255, 0, 0],
      greatCircle: true,
      getHeight: 0,
      getWidth: 2,
    }),
  ],
});

Map.current?.addLayer(
  new MapboxLayer({ id: "deckGl", deck: DeckGl.current })
);
```
Mapbox hove rhandlers
```
const hoverEnterHandler = () => {
    Map.current!.getCanvas().style.cursor = "pointer";
};
const hoverExitHandler = () => {
  Map.current!.getCanvas().style.cursor = "grab";
};

Map.current?.on("mousemove", C.REFINERY_ICON_LAYER_ID, hoverEnterHandler);
Map.current?.on("mouseleave", C.REFINERY_ICON_LAYER_ID, hoverExitHandler);
```

**Reply by Pessimistress** [reply]
Deck's `getCursor` sets `canvas.style.cursor` on every render. You can either move your logic into `getCursor`:

```js
let cursor;
const hoverEnterHandler = () => {
    cursor = "pointer";
};
const hoverExitHandler = () => {
  cursor = "grab";
};
new Deck({
  ...
  getCursor: () => cursor
});
```

Or you can manually set the style of a parent component:

```js

Map.current.getContainer().style.cursor = "pointer";

new Deck({
  ...
  getCursor: () => 'inherit'
});

```

**Reply by GarrettCannon** [reply]
That first solution worked,
Thanks a ton for the help!

**Comment by hellorayza** [comment]
Why my Mapbox-GL mouseover was not triggered...
Here is [mini repo](https://codesandbox.io/s/hardcore-yonath-c92fj?file=/src/components/HelloWorld.vue). Thank you anyway!
![image](https://user-images.githubusercontent.com/28865896/142880065-2db6d56a-2f88-4452-8791-4e4970f460ad.png)

**Reply by Pessimistress** [reply]
Mapbox does not trigger events on custom layers. Use the `onHover` prop in your deck.gl layer.
